ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_15/EventQueryServiceIT.scala

```diff
@@ -0,0 +1,269 @@
+  // For the TextKey template the key is: (tkParty, tkKey) : (Party, Text)
+  // When populating the Record identifiers are not required.
+  private def makeTextKeyKey(party: ApiTypes.Party, keyText: String) = {
+    Value(
+      Value.Sum.Record(
+        Record(fields =
+          Vector(
+            RecordField(value = Some(Value(Value.Sum.Party(party.unwrap)))),
+            RecordField(value = Some(Value(Value.Sum.Text(keyText)))),
+          )
+        )
+      )
+    )
+  }
+
+  test(
+    "TXEventsByContractIdBasic",
+    "Expose a create event by contract id",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    for {
+      tx <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, Dummy(party).create.command)
+      )
+      expected = assertDefined(
+        tx.transaction.flatMap(_.events.flatMap(_.event.created).headOption),
+        "Expected a created event",
+      )
+      events <- ledger.getEventsByContractId(
+        GetEventsByContractIdRequest(expected.contractId, Tag.unsubst(immutable.Seq(party)))
+      )
+    } yield {
+      val actual = assertDefined(events.createEvent, "Expected a created event")
+      assertEquals("Looked up event should match the transaction event", actual, expected)
+    }
+  })
+
+  test(
+    "TXEventsByContractIdConsumed",
+    "Expose an archive event by contract id",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    for {
+      dummyCid <- ledger.create(party, Dummy(party))
+      tx <- ledger.submitAndWaitForTransaction(
+        ledger.submitAndWaitRequest(party, dummyCid.exerciseDummyChoice1().command)
+      )
+      expected = assertDefined(
+        tx.getTransaction.events.flatMap(_.event.archived).headOption,
+        "Expected an exercised event",
+      )
+      events <- ledger.getEventsByContractId(
+        GetEventsByContractIdRequest(dummyCid.unwrap, immutable.Seq(party.unwrap))
+      )
+    } yield {
+      val actual = assertDefined(events.archiveEvent, "Expected a exercise event")
+      assertEquals("Looked up event should match the transaction event", actual, expected)
+    }
+  })
+
+  test(
+    "TXEventsByContractIdNotExistent",
+    "No events are returned for a non-existent contract id",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    val nonExistentContractId = ContractId.V1.assertFromString("00" * 32 + "0001")
+    for {
+      events <- ledger.getEventsByContractId(
+        GetEventsByContractIdRequest(nonExistentContractId.coid, Tag.unsubst(immutable.Seq(party)))
+      )
+    } yield {
+      assertIsEmpty(Seq(events.createEvent, events.archiveEvent).flatten[GeneratedMessage])
+    }
+  })
+
+  test(
+    "TXEventsByContractIdNotVisible",
+    "No events are returned for a non-visible contract id",
```

Comment:
We also need a test that ascertains that only a stakeholder can get contract events. Easiest version of it is some sort of ownership transfer to a new stakeholder.